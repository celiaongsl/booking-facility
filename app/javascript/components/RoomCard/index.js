import React, { Fragment } from "react";
import { useHistory } from "react-router-dom";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";

const useStyles = makeStyles({
  root: {
    maxWidth: 345,
  },
  media: {
    height: 140,
  },
  iconDiv: {
    display: "flex",
    justifyContent: "space-between",
    alignItems: "center",
    padding: "10px 0px",
    width: "40%",
    height: 44,
  },
  buttonRoot: {
    width: "100%",
  },
});

const RoomCard = (props) => {
  const { name, capacity, floor, slug } = props.room.attributes;
  const { imageURL, iconList, handleClickOpen } = props;
  const classes = useStyles();
  const history = useHistory();

  const pushToNextRoom = () => {
    history.push(`/room/${slug}`);
  };

  return (
    <Card className={classes.root}>
      <CardActionArea>
        <CardMedia className={classes.media} image={imageURL} title={name} />
        <CardContent onClick={pushToNextRoom}>
          <Typography gutterBottom variant="h5" component="h2">
            {name}
          </Typography>
          <Typography>{capacity} pax</Typography>
          <Typography variant="body2">Floor {floor}</Typography>
          <div className={classes.iconDiv}>
            {iconList &&
              iconList.map((icon, index) => (
                <Fragment key={index}>{icon}</Fragment>
              ))}
          </div>
        </CardContent>
      </CardActionArea>
      <CardActions disableSpacing>
        <Button
          size="large"
          color="primary"
          className={classes.buttonRoot}
          onClick={handleClickOpen}
        >
          Book Now
        </Button>
      </CardActions>
    </Card>
  );
};

export default RoomCard;
