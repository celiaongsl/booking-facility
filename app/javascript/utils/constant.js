import React from 'react';

import TvRoundedIcon from "@material-ui/icons/TvRounded";
import PowerRoundedIcon from '@material-ui/icons/PowerRounded';
import NoteOutlinedIcon from '@material-ui/icons/NoteOutlined';
import CastRoundedIcon from '@material-ui/icons/CastRounded';

export const assetToIconList = {
  tv: {
    icon: <TvRoundedIcon />,
    label: "TV",
  },
  power_outlets: {
    label: "Power Outlets",
    icon: <PowerRoundedIcon />
  },
  whiteboard: {
    label: "Whiteboard",
    icon: <NoteOutlinedIcon />
  },
  projector: {
    label: "Projector",
    icon: <CastRoundedIcon />
  },
};

export const apiURL = '/api/v1';