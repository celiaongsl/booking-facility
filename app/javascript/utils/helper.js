import { assetToIconList } from "./constant";

export const findAssetKeyReturnIconList = (assets) => {
  let assetsIconList = [];
  Object.keys(assets).map((key) => {
    if (assets[key]) {
      return assetsIconList.push(assetToIconList[key].icon);
    }
    return null;
  });
  return assetsIconList;
};

export const roundTimeQuarterHour = (time) => {
  var timeToReturn = new Date(time);

  timeToReturn.setMilliseconds(
    Math.round(timeToReturn.getMilliseconds() / 1000) * 1000
  );
  timeToReturn.setSeconds(Math.round(timeToReturn.getSeconds() / 60) * 60);
  timeToReturn.setMinutes(Math.round(timeToReturn.getMinutes() / 15) * 15);
  return timeToReturn;
};

// Take "Thu Feb 18 2021 19:15:00 GMT+0800 (Singapore Standard Time)"
// And shorten to "19:15:00" only
export function prettyDate2(time) {
  if (typeof time === "string") time = new Date(time);
  var localeSpecificTime = time.toLocaleTimeString();
  return localeSpecificTime.replace(/:\d+ /, " ");
}

export const convertObjectIntoQueryString = ({ object, paramKey }) => {
  // if capacity[key] === true, we add capacity[]=key in the query
  // so if we have multiple capacity keys that are true, we need to chain it as such:
  // ?capacity[]=1&capacity[]=30 << return this back

  let queryString = "";
  Object.keys(object).map((key) => {
    if (object[key]) {
      queryString = queryString + `&${paramKey}=${key}`;
    }
  });

  return queryString;
};
