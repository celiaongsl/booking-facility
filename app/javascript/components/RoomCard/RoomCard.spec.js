import React from "react";
import RoomCard from "./";
import { render } from "@testing-library/react";

// import SleepingPodImage from "../../assets/images/sleeping_pod.jpg";
import { assetToIconList, roomTypeToImageList } from "../../utils/constant";
import { findAssetKeyReturnIconList } from "../../utils/helper";
const room = {
  attributes: {
    name: "Sleeping Pod 1",
    capacity: "30",
    floor: "8",
    slug: "sleeping-pod-1",
  },
};
const imageURL = roomTypeToImageList.sleeping_pod;
const iconList = findAssetKeyReturnIconList(assetToIconList);

describe("<RoomCard />", () => {
  it("renders the component", () => {
    const { asFragment } = render(
      <RoomCard room={room} imageUrl={imageURL} iconList={iconList} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
