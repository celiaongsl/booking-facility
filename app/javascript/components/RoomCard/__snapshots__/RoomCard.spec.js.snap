// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomCard /> renders the component 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <button
      class="MuiButtonBase-root MuiCardActionArea-root"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiCardMedia-root makeStyles-media-2"
        title="Sleeping Pod 1"
      />
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          Sleeping Pod 1
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          30 pax
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Floor 8
        </p>
        <div
          class="makeStyles-iconDiv-3"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 3H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h5v1c0 .55.45 1 1 1h6c.55 0 1-.45 1-1v-1h5c1.1 0 1.99-.9 1.99-2L23 5c0-1.1-.9-2-2-2zm-1 14H4c-.55 0-1-.45-1-1V6c0-.55.45-1 1-1h16c.55 0 1 .45 1 1v10c0 .55-.45 1-1 1z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.01 7L16 4c0-.55-.45-1-1-1s-1 .45-1 1v3h-4V4c0-.55-.45-1-1-1s-1 .45-1 1v3h-.01C6.9 7 6 7.9 6 8.99v4.66c0 .53.21 1.04.58 1.41L9.5 18v2c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-2l2.92-2.92c.37-.38.58-.89.58-1.42V8.99C18 7.89 17.11 7 16.01 7z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 4H4c-1.1 0-2 .9-2 2v12.01c0 1.1.9 1.99 2 1.99h16c1.1 0 2-.9 2-2v-8l-6-6zM4 18.01V6h11v5h5v7.01H4z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 3H3c-1.1 0-2 .9-2 2v2c0 .55.45 1 1 1s1-.45 1-1V6c0-.55.45-1 1-1h16c.55 0 1 .45 1 1v12c0 .55-.45 1-1 1h-5c-.55 0-1 .45-1 1s.45 1 1 1h6c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM2.14 14.09c-.6-.1-1.14.39-1.14 1 0 .49.36.9.85.98 2.08.36 3.72 2 4.08 4.08.08.49.49.85.98.85.61 0 1.09-.54 1-1.14-.48-2.96-2.82-5.29-5.77-5.77zM1 18v3h3c0-1.66-1.34-3-3-3zm1.1-7.95c-.59-.05-1.1.41-1.1 1 0 .51.38.94.88.99 4.27.41 7.67 3.81 8.08 8.08.05.5.48.87.99.87.6 0 1.06-.52 1-1.11-.53-5.19-4.66-9.31-9.85-9.83z"
            />
          </svg>
        </div>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-buttonRoot-4 MuiButton-textPrimary MuiButton-textSizeLarge MuiButton-sizeLarge"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Book Now
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
